merge clog and fix flag handling

tbg opened the pull request:
decided to merge clog into util/log since the flag handling was ridiculous
and we were only using clog as an engine for efficient writing of log output
anyways. Unexported a bunch of things and added in code that allows the use
of log flags for the tests. This meant adding a few lines of boilerplate to
each package, but it doesn't seem too bad. Also fixed up the code for `make check`.

Also changed the default of --logtostderr to true. If one wants to log to
files only, --log-dir=/some/dir and --logtostderr=false (and presumably
--stderrthreshold=<some_number> or --alsologtostderr should also be specified).

Since that's what we'll use for machine-based logging, I'll rearrange and possibly
consolidate some of the flags in a follow-up PR which adds a "nice" (but possibly
less performant) output formatter that is intended for human consumption on a TTY.

tschottdorf commented on util/log/clog.go:
```diff
@@ -0,0 +1,1024 @@
+// Go support for leveled logs, analogous to https://code.google.com/p/google-clog/
```
mostly import from clog with the exception of `InitFlags` in this file (and a bunch of renames to unexport fields and smaller changes)

tschottdorf replied:
There's a bunch of dead code here now. I'll let it be for now and let an automated tool deal with it at some point.

tschottdorf commented on util/log/flag.go:
```diff
@@ -0,0 +1,20 @@
+// Copyright 2015 The Cockroach Authors.
```
deleting this.

tschottdorf commented on util/log/clog_file.go:
```diff
@@ -0,0 +1,125 @@
+//
+// Copyright 2013 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
```
this file is pretty much just a copy. Might have moved a variable or two but nothing else.

tbg committed:
merge clog and fix flag handling

decided to merge clog into util/log since the flag handling was ridiculous
and we were only using clog as an engine for efficient writing of log output
anyways. Unexported a bunch of things and added in code that allows the use
of log flags for the tests. This meant adding a few lines of boilerplate to
each package, but it doesn't seem too bad. Also fixed up the clog code for
`make check`.

Also changed the default of --logtostderr to true. If one wants to log to
files only, --log-dir=/some/dir and --logtostderr=false (and presumably
--stderrthreshold=<some_number> should also be specified).

petermattis commented on gossip/main_test.go:
```diff
@@ -0,0 +1,24 @@
+import "github.com/cockroachdb/cockroach/util/log"
+
+func init() {
+	log.SetupTestingFlags()
```
This sucks. So easy to forget to add this to a package.

petermattis commented on main.go:
```diff
@@ -29,24 +28,6 @@ import (
-		}
-	}
-}
-
```
Yay!

petermattis commented on util/log/flags.go:
```diff
@@ -0,0 +1,70 @@
+	pf.Var(&pflagValue{&logging.traceLocation}, "log-backtrace-at", "when logging hits line file:N, emit a stack trace")
+	logDir = pf.String("log-dir", "", "if non-empty, write log files in this directory") // in clog.go
+
+	pf.BoolVar(&humanLogging, "log-human", false, "log human-friendly output to stderr (slower)") // in log.go
```
I think you should default this value to whether stdout is a tty.

tschottdorf replied:
won't that be confusing? That means setting --logtostderr to the opposite by default, and it means that when you look at the help you see values that depend on your way of looking at the help.

petermattis replied:
I'd leave the `--logtostderr` default unchanged. Another alternative is to make this a string (or custom type) and default the value to `"auto"`.

tschottdorf replied:
If you leave it unchanged then on a TTY you get both structured and human log. Why would you want that?

petermattis replied:
Eh? Perhaps I'm misunderstanding this flag. I was imagining that `--log-human` would control the format of what is output to stderr and `--logtostderr` would control whether anything is written to stderr or not.

tschottdorf replied:
`--logtostderr` is internal to clog and turns off logging to files (but spits everything to stderr instead). I'd like to try not to muddle up glog more than we already have. Let's discuss this after this lands, I'm planning to reshuffle the flags anyways.

bdarnell replied:
How does this flag interact with `--logtostderr`? Does it write human-friendly output to stderr in addition to whatever else would be written, or does it replace the normal output? Can i have human-friendly output to stderr and machine-friendly output to files?

Note that when the tests run in circleci, stderr is not a tty, but we still want human-readable output. I don't think we should base the defaults on `isatty()`.

petermattis replied:
I'm not sure why basing the default on `isatty()` is problematic, but perhaps we should get clarity on why we want this flag at all. When logging to stderr would you ever want the machine-friendly format?

bdarnell replied:
In docker, the best practice is to not write to log files, but to write to stderr and let the docker daemon pass that to a file (and handle rotation). In that situation we would want machine-readable output on stderr. I think that defaulting to human-friendly output for stderr (regardless of tty status) is a good idea, though; we can pass a special flag in the docker configuration to request machine-readble stderr.

tschottdorf replied:
I'll address this down the road. The flags will need some general shaping up, I just stubbed out basics here.

petermattis commented on util/log/flags.go:
```diff
@@ -0,0 +1,70 @@
+	if err := pflag.CommandLine.Parse(ourArgs); err != nil {
+		panic(err.Error())
+	}
+}
```
I think it would be better for this library to install stdlib flags by default which would remove the need for `SetupTestingFlags` and `trimTestArgs`. I might actually remove `InitFlags` as well and just keep the previous code in `server/cli` for bridging `"flags"` into `"pflags"`.

tschottdorf replied:
done.

bdarnell commented on Makefile:
```diff
@@ -40,7 +40,7 @@ RACETIMEOUT  := 5m
 # STANDARDTESTFLAGS contains flags that you probably don't want to override;
 # TESTFLAGS defaults to empty so setting it doesn't clobber anything standard.
-STANDARDTESTFLAGS := -logtostderr
+STANDARDTESTFLAGS :=
```
`STANDARDTESTFLAGS` exists solely so it can pass in `-logtostderr`, so if we're changing the default of that flag we can get rid of `STANDARDTESTFLAGS`

tschottdorf replied:
done.

bdarnell commented on util/log/log.go:
```diff
@@ -23,16 +23,20 @@ import (
 )
 
-const timeFormat = "1/2 15:04:05.000" // m/d h:m:s.nanos
+const humanTimeFormat = "1/2 15:04:05.00" // m/d h:m:s.ss
```
I'd rather keep three fractional digits - I can think in milliseconds but centiseconds just look weird.

tschottdorf replied:
ok.

tbg committed:
address feedback

tschottdorf commented:
PTAL @petermattis.

petermattis commented on circle.yml:
```diff
@@ -37,8 +37,8 @@ test:
-    - docker run "cockroachdb/cockroach-dev" test TESTTIMEOUT=30s TESTFLAGS='-v -vmodule=multiraft=5,raft=1' > "${CIRCLE_ARTIFACTS}/test.log"
-    - docker run "cockroachdb/cockroach-dev" testrace RACETIMEOUT=5m TESTFLAGS='-v -vmodule=multiraft=5,raft=1' > "${CIRCLE_ARTIFACTS}/testrace.log"
+    - docker run "cockroachdb/cockroach-dev" test TESTTIMEOUT=30s TESTFLAGS='-v --vmodule=multiraft=5,raft=1' > "${CIRCLE_ARTIFACTS}/test.log"
+    - docker run "cockroachdb/cockroach-dev" testrace RACETIMEOUT=5m TESTFLAGS='-v --vmodule=multiraft=5,raft=1' > "${CIRCLE_ARTIFACTS}/testrace.log"
```
This part of the change should be unnecessary now, correct?

tschottdorf replied:
Yeah. Unnecessary, but I prefer if we refer to "our" flags homogeneously.

petermattis replied:
Yep, that's fine by me.

petermattis commented on server/cli/flags_test.go:
```diff
@@ -18,25 +18,10 @@
-		}
-	})
-}
-
```
I think we want to keep this test around.

tschottdorf replied:
whoop, adding it back.

tamird replied:
did this never get added back?

tschottdorf replied:
Looks like I checked out the wrong file. Good catch, I'll put it back in.

petermattis commented on util/log/doc.go:
```diff
@@ -0,0 +1,75 @@
+//		Logs are written to standard error instead of to files.
+//	--alsologtostderr=false
+//		Logs are written to standard error as well as to files.
+//	--stderrthreshold=ERROR
```
This flag currently doesn't exist.

tschottdorf replied:
I'll rip this out with the next PR.

tschottdorf replied:
done.

petermattis commented:
LGTM

tbg committed:
drop STANDARDTESTFLAGS from Makefile
